File: Difficulty/A. Easy/0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.c
#include <stdio.h>
#include <string.h>

/**
 * @brief Finds the index of the first occurrence of needle in haystack.
 *
 * @param haystack (char*): The main string in which to search.
 * @param needle (char*): The substring to find within haystack.
 *
 * @return (int): The index of the first occurrence of needle in haystack,
 *                or -1 if needle is not found.
 */
int strStr(char* haystack, char* needle) {
    // Get the lengths of haystack and needle
    int haystack_len = strlen(haystack);
    int needle_len = strlen(needle);
    
    // If needle is empty, return 0 as per the problem's typical definition.
    if (needle_len == 0) {
        return 0;
    }
    
    // Loop through the haystack, stopping when there aren't enough characters
    // left to match the needle.
    for (int i = 0; i <= haystack_len - needle_len; i++) {
        // Check if the substring starting at position i matches needle.
        int j;
        for (j = 0; j < needle_len; j++) {
            if (haystack[i + j] != needle[j]) {
                break;  // Mismatch found, break out of the inner loop
            }
        }
        
        // If we've gone through the entire needle without a mismatch, return i.
        if (j == needle_len) {
            return i;
        }
    }
    
    // If needle was not found in haystack, return -1.
    return -1;
}

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        """
        Finds the index of the first occurrence of needle in haystack,
        or returns -1 if needle is not part of haystack.

        Parameters:
        haystack (str): The main string where we search for the needle.
        needle (str): The substring we are looking for within haystack.

        Returns:
        int: The starting index of the first occurrence of needle in haystack,
             or -1 if needle is not found.
        """
        
        # Use the built-in find() method to locate the first occurrence
        # of needle in haystack. This method returns the index if found,
        # or -1 if the substring is not present in haystack.
        return haystack.find(needle)
